<html>
<head>
<meta name="order" content="2" />
<title>Microarray probeset mapping</title>
</head>

<body>

<h1>Microarray probeset mapping</h1>

<h2>Inclusion of microarray data</h2>

<p>Microarray data are added to Ensembl Genomes after consultation with relevant scientific 
user communities to identify experiments of general interest within that community 
(e.g. <a href="http://www.wormbase.org">WormBase</a> for worms, <a href="http://www.vectorbase.org">VectorBase</a> for arthropod disease vectors). Suggestions for new data sets are welcome, and 
wherever possible these will be derived from publically accessible data sources such as 
<a href="http://www.ebi.ac.uk/arrayexpress/">Array Express</a>.</p>

<h2>Microarray probe mapping</h2>

<p>Microarray probe mapping is conducted for non-VectorBase species as described by the 
<a href="/info/genome/microarray_probe_set_mapping.html">Ensembl microarray probe mapping pipeline</a>. The following modifications have been made for VectorBase-derived species.</p>

<h3>Step one: Genome/transcript alignment</h3>

<p>Probes are mapped using two different programs according to probe length - short probes 
(Bowtie2 alignment, while longer probes (&gt;200bp) are mapped using <a href="http://www.biomedcentral.com/1471-2105/6/31/abstract">Exonerate</a>. The reason for processing the probes differently 
according to length is based on the need to rapidly map large numers of probes across many 
different species. The majority of VectorBase probes are short (Bowtie2 alignments), which are 
significantly faster to perform than Exonerate. Longer probes (&gt;200bp) are often derived from 
cDNA/EST sequences and span multiple exon/intron junctions - these probes are mapped with 
Exonerate as it detects these junctions in an accurate and sensitive manner. Probes are mapped 
against both the genomic sequence and the transcripts. Where high quality hits are found on the 
transcript but not on the genomic sequence, the probe sequence is projected from the transcript 
sequence back into genomic coordinates.</p>

<p>Alignments are stored as described by the Ensembl microarray probe mapping pipeline, but only 
probes that match 10 or fewer unique locations in the genome are stored.</p>

<h3>Step two: Ensembl transcript annotation</h3>

<p>The association of probes and probe sets with transcripts is performed as described in the 
Ensembl microarray probe mapping pipeline, except that for VectorBase species the UTR annotation 
of the transcript is followed strictly and no attempt is made to extend the UTR region based on 
the mean/modal UTR statistics of the transcripts.</p>

<h2>Data access</h2>

<p>The probe mappings and transcript annotations are stored in Ensembl functional genomics (funcgen) 
databases and can be programatically accessed using the <a href="http://www.ensembl.org/info/docs/api/funcgen/index.html">funcgen API</a>. POD documentation is available:</p>

<ul>
  <li><a href="http://www.ensembl.org/info/docs/Doxygen/funcgen-api/classBio_1_1EnsEMBL_1_1Funcgen_1_1Array.html">Bio::EnsEMBL::Funcgen::Array</a></li>
  <li><a href="http://www.ensembl.org/info/docs/Doxygen/funcgen-api/classBio_1_1EnsEMBL_1_1Funcgen_1_1Probe.html">Bio::EnsEMBL::Funcgen::Probe</a></li>
  <li><a href="http://www.ensembl.org/info/docs/Doxygen/funcgen-api/classBio_1_1EnsEMBL_1_1Funcgen_1_1ProbeSet.html">Bio::EnsEMBL::Funcgen::ProbeSet</a></li>
  <li><a href="http://www.ensembl.org/info/docs/Doxygen/funcgen-api/classBio_1_1EnsEMBL_1_1Funcgen_1_1ProbeFeature.html">Bio::EnsEMBL::Funcgen::ProbeFeature</a></li>
</ul>

<p>Probe and ProbeSet level transcript annotations are stored in the funcgen databases and can be 
accessed using the API, or the relevant Ensembl Genomes BioMart.</p>

</body>
</html>
