<html>
<head>
<meta name="order" content="2" />
  <title>Comparative Genomics</title>
</head>
<body>

<h1>Comparative Genomics</h1>

<h2>Data available</h2>

<ul>
<li><strong>Gene trees</strong> are constructed using the longest protein for every gene in EnsemblGenomes.
<strong>Homologues</strong> are deduced from these trees. Proteins are clustered based on Best-Reciprocal Hits
and Blast Score Ratios, and each cluster of proteins is aligned using Muscle. Finally, TreeBeST is used to produce
a gene tree from each multiple alignment, reconciling it with the species tree to call duplication events.
<a href="homology_method.html">More information</a></li>

<li><strong>Whole genome alignments</strong> are performed using multiple species. <a href="analyses.html">More information &rarr;</a></li>

<li><strong>Ancestral sequences</strong> are calculated from multi-species whole genome
alignments.  <a href="analyses.html#ancestral">More information&rarr;</a></li>

<li><strong> Conservation scores and constrained elements </strong> are
calculated from the whole genome multiple alignments.
<a href="analyses.html#conservation">More information&rarr;</a></li>

<li><strong> Syntenies </strong> are calculated from the multiple alignments.
<a href="analyses.html#synteny">More information&rarr;</a></li>

</ul>

<h2>Access</h2>
<p>
<strong>Data</strong> can be accessed using the <a href="/info/docs/api/compara/index.html">Compara Perl API</a>,
<a href="/info/data/biomart.html">BioMart</a>, or comparative genomics pages on the browser.
Gene trees can be viewed from any 'Gene' page on the browser, and exported via the control panel.
</p>

<h2>Species Trees</h2>

<h3>Clade Compara</h3>

[[INCLUDE::/info/genome/compara/cladetree.inc]]

<h3>Pan-taxonomic Compara</h3>

[[INCLUDE::/info/genome/compara/pan.inc]]

</body>
</html>
