<html>
<head>
<meta name="order" content="2" />
<title>Pan-taxonomic compara</title>
</head>

<body>

<h1>Pan-taxonomic compara</h1>

<p>The pan-taxonomic compara is built using representative genomes from all significant 
clades represented in <a href="http://www.ensembl.org/">Ensembl</a> and 
<a href="http://www.ensemblgenomes.org/">Ensembl Genomes</a>, to offer a broad view of homologous 
relationships from across the taxonomy. This analysis enables the production of protein trees 
and orthologies across all the taxonomic groups in Ensmbl and Ensembl Genomes. 

<h2>Species in the pan-taxonomic compara</h2>

<!--<p>A complete list of all genomes included can be found <a href="http://ensemblgenomes.org/info/genomes?name=&amp;division=All&amp;pan_compara=1&amp;variation=All&amp;field_peptide_alignments_value=All&amp;field_genome_alignments_value=All&amp;field_other_alignments_value=All">here</a>-->.

<p>Model genomes along with genomes not represented by the familes of model organisms are picked for this analysis.</p>

<p>The bacterial genomes included are selected on the following criteria:</p>

<ol>
  <li>Presence in earlier versions of the pan-taxonomic compara</li>
  <li>Presence in the UniProtKB <a href="http://www.uniprot.org/taxonomy/complete-proteomes">reference proteome list</a></li>
  <li><a href="http://europepmc.org/">Europe PubMed Central</a> citation count of over 1000</li>
</ol>

<h2>Data Access </h2>

<p>The gene trees obtained from this analysis can be accessed in the <a href="/info/website/tutorials/compara.html">genome browser</a> and also using the <a href="/info/data/api.html">Perl API</a> 
and <a href="http://rest.ensembl.org">REST API</a>. Note that when using the APIs, 
the pan-taxonomic comparative genomics data resource can be picked using the name "pan_homology".</p>


</body>
</html>
