<html>
<head>
<meta name="order" content="4" />
<title>Synteny</title>
</head>

<body>

<h1>Synteny</h1>

<p>The method of calculating synteny in Ensembl Genomes depends on the division, to account for 
differences in the underlying data such as genome size and evolutionary distance.</p>

<h2>Ensembl Fungi and Ensembl Protists</h2>

<p>Synteny is calculated using a modified form of the algorithm as specified in 
<a href="http://www.ensembl.org/info/genome/compara/analyses.html#synteny">Ensembl</a>, 
based on <a href="/info/data/whole_genome_alignment">pairwise whole genome alignments</a>. 
The synteny algorithm has two phases. First, orthologous regions that are closer than 1.5kbp are 
grouped. In the second phase, the regions that are in synteny are linked provided that no more 
than 2 non-syntenic groups are found between them, and they are less than 3kbp apart.</p>

<p>Current synteny data is available from:</p>

<ul>
  <li><a href="http://protists.ensembl.org/info/genome/compara/compara_analyses.html">Ensembl Protists</a></li>
  <li><a href="http://fungi.ensembl.org/info/genome/compara/compara_analyses.html">Ensembl Fungi</a></li>
</ul>

<h2>Ensembl Metazoa</h2>

<p>Nematode synteny is based on <a href="/info/genome/compara/whole_genome_alignment.html">pairwise whole genome alignments</a>; Dipteran synteny is based on <a href="/info/genome/compara/peptide_compara.html">orthology</a>. In the latter case, all orthologues between the species were used to generate 
syntenic blocks, with the restriction that at least two orthologs were present in each block. Both sets of synteny use the standard <a href="http://www.ensembl.org/info/genome/compara/analyses.html#synteny">Ensembl</a> pipeline parameters and a two-stage synteny algorithm. First, orthologous regions that are closer than 200kbp are grouped. In the second stage, the regions that are in synteny are linked provided that no more than 2 non-syntenic groups are found between them, and they are less than 3Mbp apart.</p>

<p>Current synteny data is available from <a href="http://metazoa.ensembl.org/info/genome/compara/compara_analyses.html">Ensembl Metazoa</a>.</p>

<h2>Ensembl Plants</h2>

<p>Ensembl Plants includes synteny analyses generated in house, based on <a href="http://www.ensembl.org/info/genome/compara/synteny.html">pairwise genome alignments</a>, and some generated by our collaborators <a href="http://gramene.org/">Gramene</a>. The Gramene method is based on <a href="/info/genome/compara/peptide_compara.html">orthology</a>. 
For each species combination, the orthologues are assigned a gene index (1..last) depending on order along the chromosome (non-orthologous genes are skipped). DAGchainer software is used to detect collinear genes contained in syntenic blocks, and the coordinates of the syntenic block are derived from the outermost genes within each block. There is no post-processing, so syntenic blocks can sometimes overlap or be nested. The method was used in <a href="http://europepmc.org/abstract/MED/19965430">The B73 Maize Genome: Complexity, Diversity, and Dynamics</a>. Schnable PS <i>et al</i>. 2009. Science 326(5956):1112-1115.</p>

<p>Current synteny data is available from <a href="http://plants.ensembl.org/info/genome/compara/compara_analyses.html">Ensembl Plants</a>.</p>

</body>
</html>
