<html>
<head>
<meta name="order" content="3" />
<title>Whole genome alignment</title>
</head>

<body>

<h1>Whole genome alignment</h1>

<p>Two types of pairwise genome alignment are available in Ensembl Genomes, based on LastZ [1] 
(or its predecessor BlastZ [2]) and translated BLAT (tBLAT) [3]. LastZ is typically used for 
closely related species, and tBLAT for more distant species. The method of alignment affects 
the coverage of the genomes, with tBLAT expected to mostly find homologies in coding regions.</p>

<p>The raw results from LastZ or tBLAT are alignment blocks, which are 'chained' according to their 
location in both genomes, then 'netted' to choose the best sub-chain in each region for the 
reference species [4]. The resultant <b>LastZ-net</b> and <b>tBLAT-net</b> alignments are displayed 
in Ensembl Genomes for selected <a href="http://fungi.ensembl.org/info/genome/compara/compara_analyses.html">fungal</a>, 
<a href="http://metazoa.ensembl.org/info/genome/compara/compara_analyses.html">metazoan</a>, 
<a href="http://protists.ensembl.org/info/genome/compara/compara_analyses.html">protist</a> and 
<a href="http://plants.ensembl.org/info/genome/compara/compara_analyses.html">plant</a> species.</p>

<h2>References</h2>

<ol>
  <li><a href="http://www.bx.psu.edu/~rsharris/lastz">Improved pairwise alignment of genomic DNA</a>. Harris RS. 2007. Ph.D. Thesis, The Pennsylvania State University.</li>
  <li><a href="http://europepmc.org/abstract/MED/12529312">Human-mouse alignments with BLASTZ</a>. Schwartz S <em>et al</em>. 2003. Genome Research. 13(1):103-7.</li>
  <li><a href="http://europepmc.org/abstract/MED/11932250">BLAT--the BLAST-like alignment tool</a>. Kent WJ. 2002. Genome Research. 12(4):656-64.</li>
  <li><a href="http://europepmc.org/abstract/MED/14500911">Evolution's cauldron: duplication, deletion, and rearrangement in the mouse and human genomes</a>. Kent WJ <em>et al</em>. 2003. PNAS. 100(20):11484-9.</li>
</ol>

</body>
</html>
