<html>
<head>
<meta name="order" content="1" />
<title>Variation</title>
</head>

<body>

<h1>[[INCLUDE::/info/genome/variation/site_var_heading.inc]]</h1>

<p>
Ensembl Genomes imports variations including Single Nucleotide Polymorphisms
(SNPs) and insertion-deletion mutations (Indels) from various sources. 
Any effects on transcripts in the area are calculated. 
</p>

[[INCLUDE::/info/genome/variation/site_var.inc]]

<p>
Variations and sources can be viewed in the browser through pages such
as:
</p>

<ul>
<li><strong>Gene:</strong> Variation table and Variation image (for
all variations in a gene)</li>
<li><strong>Transcript:</strong> Population comparison and Comparison
image (for all variations in a transcript across different
sequences)</li>
<li><strong>Location:</strong> Region in Detail (Variations can be
drawn using &quot;Configure this page&quot; at the left). The menu allows
display of information in Ensembl databases along with external
sources in DAS format.
</ul>

<p><strong>Variation Types</strong></p>

<p>
The variation types describe the location or effect on the transcript.
</p>

<ul>

<li><strong>Non-synonymous SNPs</strong> - SNPs that are located in
the coding sequence and result in an amino acid change in the encoded
peptide sequence.</li>

<li><strong>Synonymous SNPs</strong> - In coding sequence, not
resulting in an amino acid change (i.e. silent mutation).</li>

<li><strong>Frameshift variations</strong> - In coding sequence,
resulting in a frameshift.</li>

<li><strong>Stop lost</strong> - In coding sequence, resulting in the
loss of a stop codon.</li>

<li><strong>Stop gained</strong> - In coding sequence, resulting in
the gain of a stop codon (i.e. leading to a shortened peptide
sequence).</li>

<li><strong>Essential splice site</strong> - In the first 2 or the
last 2 basepairs of an intron.</li>

<li><strong>Splice site</strong> - 1-3 bps into an exon or 3-8 bps
into an intron.</li>

<li><strong>Upstream variations</strong> - Within 5 kb upstream of the
5prime end of a transcript.</li>

<li><strong>Regulatory region variations</strong> - In regulatory
region annotated by Ensembl.</li>

<li><strong>5prime UTR variations</strong> - In 5prime UTR
(untranslated region).</li>

<li><strong>Intronic variations</strong> - In intron.</li>

<li><strong>3prime UTR variations</strong> - In 3prime UTR.</li>

<li><strong>Downstream variations</strong> - Within 5 kb downstream of
the 3prime end of a transcript.</li>

<li><strong>Intergenic variations</strong> - More than 5 kb either
upstream or downstream of a transcript.</li>

<li><strong>Complex InDel</strong> - Insertion or deletion that spans
an exon/intron border or a coding sequence/UTR border.</li>

<li><strong>Within mature miRNA</strong> - Located within a
microRNA.</li>

<li><strong>Within non-coding gene</strong> - Located within a gene
that does not code for a protein.</li>

</ul>

<p>
Clicking on any variation on an Ensembl page will open a Variation tab
with information about the flanking sequence and source for the
selected variation. Links to genes and
transcripts that include the variation can be found at the left of
this tab. 
</p>

<p>
Variation information can also be accessed
using <a href="/info/data/biomart.html"
class="cp-external">BioMart</a> (gene database), and
the <a href="/info/docs/api/index.html" class="cp-external">Perl
API</a> (<a href="/info/docs/api/variation/index.html"
class="cp-external">variation databases</a>).
</p>

[[INCLUDE::/info/genome/variation/site_var_ref.inc]]

</body>
</html>
