<html>
<head>
<title>Compara API Tutorial</title>
</head>

<body>

<h1>Compara API Tutorial</h1>


<h2>Introduction</h2>

<p>
This tutorial is an introduction to the <a href="index.html">EnsEMBL Compara API</a>.
Knowledge of the
<a href="/info/docs/api/core/index.html">EnsEMBL Core API</a> and of the concepts and conventions in the
<a href="/info/docs/api/core/core_tutorial.html">EnsEMBL Core API tutorial</a> is assumed.
Documentation about the Compara database schema is available at
<a href="http://cvsweb.sanger.ac.uk/cgi-bin/cvsweb.cgi/ensembl-compara/docs/">http://cvsweb.sanger.ac.uk/cgi-bin/cvsweb.cgi/ensembl-compara/docs/</a>
(or in ensembl-compara/docs/docs/schema_doc.html from the EnsEMBL CVS repository), and while is not necessary for this tutorial,
an understanding of the database tables may help as many of the adaptor modules are table-specific.
</p>


<h2>Installing the API</h2>

<p>
<a href="/info/docs/api/api_installation.html">API installation</a> and updating is the same as per the core API.
</p>

<h2>Connecting to an EnsEMBL Compara database</h2>

<h3>Connection parameters</h3>

<h3>There are two API ways to connect to the EnsEMBL Compara database:</h3>
<ul>
<li>In most cases you will prefer the implicit way - using Bio::EnsEMBL::Registry module,
which can read either a global or a specific configuration file or auto-configure itself.</li>
<li>However there are cases where you might want more flexibility provided by the
explicit creation of a Bio::EnsEMBL::Compara::DBSQL::DBAdaptor.</li>
</ul>

<h3>Implicitly, using the Bio::EnsEMBL::Registry auto-configuration feature (recommended)</h3>

<p>
For using the auto-configuration feature, you will first need to supply the connection parameters to the
Registry loader. For instance, if you want to connect to the the public EnsEMBL databases you can
use the following command in your scripts:
</p>

<pre class="code">
use strict;
use warnings;
use Bio::EnsEMBL::Registry;

Bio::EnsEMBL::Registry-&gt;load_registry_from_multiple_dbs(
{
    -HOST =&gt; 'mysql.ebi.ac.uk',
    -PORT =&gt; 4157,
    -USER =&gt; 'anonymous'
},
{
    -HOST =&gt; 'ensembldb.ensembl.org',
    -PORT =&gt; 5306,
    -USER =&gt; 'anonymous'
}
);
</pre>


<p>
This will initialize the Registry, from which you will be able to create object-specific adaptors later.

Alternatively, you can use a shorter version based on a URL:</p>

<pre class="code">
use strict;
use warnings;
use Bio::EnsEMBL::Registry;
Bio::EnsEMBL::Registry-&gt;load_registry_from_url('mysql://anonymous@mysql.ebi.ac.uk:4157/');
</pre>

<h3>Implicitly, using the Bio::EnsEMBL::Registry configuration file</h3>

<p>
You will need to have a registry configuration file set up.
By default, it takes the file defined by the ENSEMBL_REGISTRY environment variable or
the file named .ensembl_init in your home directory if the former is not found.
Additionally, you can use a specific file
(see perldoc Bio::EnsEMBL::Registry or later in this document for some examples on how to use a different file).
Please, refer to the <a href="/info/docs/api/registry.html">EnsEMBL registry documentation</a> for details about this option.
</p>

<h3>Explicitly, using the Bio::EnsEMBL::Compara::DBSQL::DBAdaptor</h3>

<p>
EnsEMBL Compara data, like core data, is stored in a MySQL relational database.
If you want to access a Compara database, you will need to connect to it.
This is done in exactly the same way as to connect to an EnsEMBL core database,
but using a Compara-specific DBAdaptor. One parameter you have to supply
in addition to the ones needed by the Registry is the <code>-dbname</code>,
which by convention contains the release number of Ensembl and Ensembl Genomes:
</p>

<pre class="code">
use strict;
use warnings;

use Bio::EnsEMBL::Compara::DBSQL::DBAdaptor;

my $host   = 'mysql.ebi.ac.uk';
my $user   = 'anonymous';
my $port   = 4157;
my $dbname = 'ensembl_compara_bacteria_6_59';

my $comparadb= new Bio::EnsEMBL::Compara::DBSQL::DBAdaptor(
    -host   =&gt; $host,
    -port   =&gt; $port,
    -user   =&gt; $user,
    -dbname =&gt; $dbname,
);
</pre>

<p>
As of release 3 of Ensembl Genomes there are 6 compara databases available;
one for each division of Ensembl Genomes and one which contains pan-taxonomic compara.
If you use the registry to configure your available adaptors you can access
each using the following names:
</p>

<ul>
<li>bacteria</li>
<li>fungi</li>
<li>metazoa</li>
<li>pan_homology</li>
<li>plants</li>
<li>protists</li>
</ul>

<h2>EnsEMBL Compara object-specific adaptors</h2>

<p>
EnsEMBL Compara adaptors are used to fetch data from the database. Data are returned as EnsEMBL objects.
For instance, the GenomeDBAdaptor returns Bio::EnsEMBL::Compara::GenomeDB objects.
</p>

<p>
Below is a non-exhaustive list of EnsEMBL Compara adaptors that are most often used:
</p>

<ul>
<li>GenomeDBAdaptor to fetch Bio::EnsEMBL::Compara::GenomeDB objects</li>
<li>DnaFragAdaptor to fetch Bio::EnsEMBL::Compara::DnaFrag objects</li>
<li>GenomicAlignBlockAdaptor to fetch Bio::EnsEMBL::Compara::GenomicAlignBlock objects</li>
<li>DnaAlignFeatureAdaptor to fetch Bio::EnsEMBL::DnaDnaAlignFeature objects
(note that this adaptor returns an EnsEMBL Core object)</li>
<li>SyntenyAdaptor to fetch Bio::EnsEMBL::Compara::SyntenyRegion objects</li>
<li>MemberAdaptor to fetch Bio::EnsEMBL::Compara::Member objects</li>
<li>HomologyAdaptor to fetch Bio::EnsEMBL::Compara::Homology objects</li>
<li>PeptideAlignFeatureAdaptor to fetch Bio::EnsEMBL::Compara::PeptideAlignFeature objects</li>
</ul>

<p>
Only some of these adaptors will be used for illustration as part of this tutorial through commented perl scripts code.
</p>

<p>
You can get the adaptors from the Registry with the get_adaptor command. You need to specify three arguments: the
species name, the type of database and the type of object. Therefore, in order to get the GenomeDBAdaptor for the
Compara database, you will need the following command:
</p>

<pre class="code">
my $genome_db_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    'pan_homology', 'compara', 'GenomeDB');

</pre>

<p>
Or an alternative way to get adaptors from a single DBAdaptor:
</p>

<pre class='code'>
my $dba = Bio::EnsEMBL::Registry-&gt;get_DBAdaptor(-species =&gt; 'pan_homology', -group =&gt; 'compara');
my $genome_db_adaptor = $dba-&gt;get_GenomeDBAdaptor();
</pre>

<p>
<strong>NB:</strong> As the EnsEMBL Compara DB is a multi-species database resource, a possible name is 'pan_homology'. The type of the
database is 'compara'. Other resources are available; see earlier on for more information of the types available.
</p>

<h2>Code Conventions</h2>

<p>
Refer to the <a href="/info/docs/api/core/index.html">EnsEMBL core tutorial</a> for a good description of the coding conventions normally used in EnsEMBL.
</p>

<p>
We can divide the fetching methods of the ObjectAdaptors into two categories: the <code>fetch_by</code> and <code>fetch_all_by</code>. The former return one single object while the latter return a reference to an array of objects.
</p>

<pre class="code">
my $this_genome_db = $genome_db_adaptor-&gt;fetch_by_name_assembly(&quot;Homo sapiens&quot;, &quot;GRCh37&quot;);
my $another_genome_db = $genome_db_adaptor-&gt;fetch_by_name_assembly(&quot;E_coli_K12&quot;, &quot;EB1&quot;);
</pre>

<pre class="code">
my $all_genome_dbs = $genome_db_adaptor-&gt;fetch_all();
foreach my $this_genome_db (@{$all_genome_dbs}) {
  print $this_genome_db-&gt;name, &quot;\n&quot;;
}
</pre>

<h2>Core Compara Objects</h2>

<h3>GenomeDB</h3>

<p>
The GenomeDB object is a representation of a species within the compara
database. It is uniqely defined in the database by a genome's name and
its assembly. In order to retrieve a GenomeDB you must use the GenomeDBAdaptor
object which has multiple methods available for finding them.
</p>

<pre class='code'>
use strict;
use warnings;
my $dba = Bio::EnsEMBL::Registry-&gt;get_DBAdaptor(-species =&gt; 'pan_homology', -group =&gt; 'compara');
my $gdba = $dba-&gt;get_GenomeDBAdaptor();

my $core_dba = get_core_dba_from_somewhere();

$gdba-&gt;fetch_by_name_assembly('e_coli_k12', 'EB1'); #Requires you to use the exact name found in compara
$gdba-&gt;fetch_by_name_assembly('e_coli_k12'); #Gets the default assembly
$gdba-&gt;fetch_by_registry_name('Escherichia coli K12'); #Allows you to use any synonym a species is found under in the registry
$gdba-&gt;fetch_by_core_DBAdaptor($core_dba); #Use a core DBAdaptor
$gdba-&gt;fetch_by_taxon_id(9606); #Fetch human by taxon
</pre>

<p>
The best methods to use when retrieving this data are
<em>fetch_by_registry_name</em> and <em>fetch_by_core_DBAdaptor</em> since both
of these methods do not assume what the name is actually known as in compara.
This is of great importance to bacterial comparas where the full name is
too long to be used in compara and contains characters which are detrimental to
the running of the pipeline (e.g. strain information). In these situations
the API is configured to use a meta item from the core database called
<em>species.compara_name</em>. The current algorithm is:
</p>

<ul>
<li>Check to see if <em>species.compara_name</em></li>
<li>If exists then use &amp; raise an error when a GenomeDB does not exist with that name</li>
<li>If not then use the binomial name from the meta adaptor</li>
<li>If does not exist then raise an error</li>
<li>Return the GenomeDB found</li>
</ul>

<p>
This promotes a level of separation between what compara uses to refer to a
species and what the ensembl code base requires a species to be known as.
</p>

<p>
Also be aware that using a taxon identifier can sometimes result in incorrect
lookups if the compara database has species which share a taxon identifier.
This happens in Ensembl Bacteria's compara database but is normally a safe
operation in most other comparas.
</p>

<h3>Method Link Species Set</h3>

<p>
These act as experimental declarations for compara. If anything has been
analysed using a compara pipeline there should be a Method Link Species Set
declaring what the algorithm/analysis used was and which species (one or
many) were invovled in the procedure. To this end there are normally three
classes of species sets:
</p>

<ol>
<li>Singletons - a species was assessed on its own e.g. paralogs</li>
<li>Pairwise - two species were assessed against each other e.g. DNA alignments</li>
<li>Groups - any number of species were assessed against each other e.g. GeneTrees</li>
</ol>

<p>
In order to find a declaration you must first know what the type of method
link you are searching for &amp; find the GenomeDBs you expect to be found in
the species set. For example:
</p>

<pre class='code'>
use strict;
use warnings;

my $dba = Bio::EnsEMBL::Registry-&gt;get_DBAdaptor(-species =&gt; 'pan_homology', -group =&gt; 'compara');
my $gdba = $dba-&gt;get_GenomeDBAdaptor();
my $mlssa = $dba-&gt;get_MethodLinkSpeciesSetAdaptor();

my $mlss = $mlssa-&gt;fetch_by_method_link_type_registry_aliases('ENSEMBL_ORTHOLOGUES',
  ['human', 'Escherichia coli K12']);

#Or
my $ss = [$gdba-&gt;fetch_by_registry_name('e_coli_k12'), $gdba-&gt;fetch_by_registry_name('human')];
$mlss = $mlssa-&gt;fetch_by_method_link_type_GenomeDBs('ENSEMBL_ORTHOLOGUES', $ss);
</pre>

<h2>Whole Genome Alignments</h2>

<p>
The Compara database contains a number of different types of whole genome alignments.
A listing about what are these different types can be found in the ensembl-compara/docs/schema_doc.html document in method_link section.
</p>

<h3>GenomicAlignBlock objects</h3>

<p>
GenomicAlignBlocks are the preferred way to store and fetch genomic alignments.
A GenomicAlignBlock contains several GenomicAlign objects.
Every GenomicAlign object corresponds to a piece of genomic sequence aligned with the other GenomicAlign in the same GenomicAlignBlock.
A GenomicAlign object is always in relation with other GenomicAlign objects
and this relation is defined through the GenomicAlignBlock object.
Therefore the usual way to fetch genomic alignments is by fetching GenomicAlignBlock objects.
We have to start by getting the corresponding adaptor:
</p>

<pre class="code">
# Getting the GenomicAlignBlock adaptor:
my $genomic_align_block_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    'bacteria', 'compara', 'GenomicAlignBlock');
</pre>

<p>
In order to fetch the right alignments we need to specify a couple of data:
the type of alignment and the piece of genomic sequence in which we are looking for alignments.
The type of alignment is a more tricky now:
you need to specify both the alignment method and the set of genomes.
In order to simply this task, you could use the new Bio::EnsEMBL::Compara::MethodLinkSpeciesSet object.
The best way to use them is by fetching them from the database:
</p>

<pre class="code">
# Getting the GenomeDB adaptor:
my $genome_db_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    'fungi', 'compara', 'GenomeDB');

# Fetching GenomeDB objects
my $cer_genome_db = $genome_db_adaptor-&gt;fetch_by_name_assembly('Saccharomyces cerevisiae');
my $pom_genome_db = $genome_db_adaptor-&gt;fetch_by_name_assembly('Schizosaccharomyces pombe');

# Getting the MethodLinkSpeciesSet adaptor:
my $method_link_species_set_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    'fungi', 'compara', 'MethodLinkSpeciesSet');

# Fetching the MethodLinkSpeciesSet object corresponding to TRANSLATED_BLAT alignments between cerevisiae and pombe genomic sequences:
my $cer_pom_blastz_net_mlss =
    $method_link_species_set_adaptor-&gt;fetch_by_method_link_type_GenomeDBs(
        &quot;TRANSLATED_BLAT&quot;,
        [$cer_genome_db, $pom_genome_db]
    );
</pre>

<p>
There are two ways to fetch GenomicAlignBlocks.
One uses Bio::EnsEMBL::Slice objects while the second one is based on
Bio::EnsEMBL::Compara::DnaFrag objects for specifying the piece of genomic
sequence in which we are looking for alignments.
</p>

<pre class="code">
my $query_species = 'Schizosaccharomyces pombe';
my $seq_region = 'I';
my $seq_region_start = 45000;
my $seq_region_end   = 47000;

# Getting the Slice adaptor:
my $slice_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    $query_species, 'core', 'Slice');

# Fetching a Slice object:
my $query_slice = $slice_adaptor-&gt;fetch_by_region(
    'toplevel',
    $seq_region,
    $seq_region_start,
    $seq_region_end);

# Fetching all the GenomicAlignBlock corresponding to this Slice:
my $genomic_align_blocks =
    $genomic_align_block_adaptor-&gt;fetch_all_by_MethodLinkSpeciesSet_Slice(
      $cer_pom_blastz_net_mlss,
      $query_slice);
</pre>

<p>
Here is an example script with all of this:
</p>

<pre class="code">
use strict;
use Bio::EnsEMBL::Registry;
use Bio::EnsEMBL::Utils::Exception qw(throw);
use Bio::SimpleAlign;
use Bio::AlignIO;
use Bio::LocatableSeq;
use Getopt::Long;

my $usage = qq{
perl DumpMultiAlign.pl
  Getting help:
    [--help]

  For the query slice:
    [--species species]
        Query species. Default is &quot;Schizosaccharomyces pombe&quot;
    [--coord_system coordinates_name]
        Query coordinate system. Default is &quot;chromosome&quot;
    --seq_region region_name
        Query region name, i.e. the chromosome name
    --seq_region_start start
    --seq_region_end end

  For the alignments:
    [--alignment_type method_link_name]
        The type of alignment. Default is &quot;TRANSLATED_BLAT&quot;
    [--set_of_species species1:species2:species3:...]
        The list of species used to get those alignments. Default is
        &quot;Saccharomyces cerevisiae:Schizosaccharomyces pombe&quot;. The names should correspond to the name of the
        core database in the registry_configuration_file or any of its
        aliases

  Ouput:
    [--output_format clustalw|fasta|...]
        The type of output you want. &quot;clustalw&quot; is the default.
    [--output_file filename]
        The name of the output file. By default the output is the
        standard output
};

my $species = &quot;Schizosaccharomyces pombe&quot;;
my $coord_system = &quot;chromosome&quot;;
my $seq_region = &quot;I&quot;;
my $seq_region_start = 45000;
my $seq_region_end = 47000;
my $alignment_type = &quot;TRANSLATED_BLAT&quot;;
my $set_of_species = &quot;Saccharomyces cerevisiae:Schizosaccharomyces pombe&quot;;
my $output_file = undef;
my $output_format = &quot;clustalw&quot;;
my $help;

GetOptions(
    &quot;help&quot; =&gt; \$help,
    &quot;species=s&quot; =&gt; \$species,
    &quot;coord_system=s&quot; =&gt; \$coord_system,
    &quot;seq_region=s&quot; =&gt; \$seq_region,
    &quot;seq_region_start=i&quot; =&gt; \$seq_region_start,
    &quot;seq_region_end=i&quot; =&gt; \$seq_region_end,
    &quot;alignment_type=s&quot; =&gt; \$alignment_type,
    &quot;set_of_species=s&quot; =&gt; \$set_of_species,
    &quot;output_format=s&quot; =&gt; \$output_format,
    &quot;output_file=s&quot; =&gt; \$output_file);

# Print Help and exit
if ($help) {
    print $usage;
    exit(0);
}

if ($output_file) {
    open(STDOUT, &quot;&gt;$output_file&quot;) or die(&quot;Cannot open $output_file&quot;);
}

my $compara = 'fungi';

Bio::EnsEMBL::Registry-&gt;load_registry_from_db(
    -host =&gt; 'mysql.ebi.ac.uk', -port =&gt; 4157, -user =&gt; 'anonymous');

# Getting all the Bio::EnsEMBL::Compara::GenomeDB objects
my $genome_dbs;
my $genome_db_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    $compara, &quot;compara&quot;, &quot;GenomeDB&quot;);

throw(&quot;Cannot connect to Compara&quot;) if (!$genome_db_adaptor);

foreach my $this_species (split(&quot;:&quot;, $set_of_species)) {
    my $this_meta_container_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
        $this_species, 'core', 'MetaContainer');

    throw(&quot;Registry configuration file has no data for connecting to &lt;$this_species&amp;&quot;)
        if (!$this_meta_container_adaptor);

    # Fetch Bio::EnsEMBL::Compara::GenomeDB object
    my $name = $genome_db_adaptor-&gt;get_species_name_from_core_MetaContainer($this_meta_container_adaptor);
    my $genome_db = $genome_db_adaptor-&gt;fetch_by_name_assembly($name);

    # Add Bio::EnsEMBL::Compara::GenomeDB object to the list
    push(@$genome_dbs, $genome_db);
}

# Getting Bio::EnsEMBL::Compara::MethodLinkSpeciesSet object
my $method_link_species_set_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    $compara, &quot;compara&quot;, &quot;MethodLinkSpeciesSet&quot;);

my $method_link_species_set =
    $method_link_species_set_adaptor-&gt;fetch_by_method_link_type_GenomeDBs(
      $alignment_type,
      $genome_dbs);

throw(&quot;The database do not contain any $alignment_type data for $set_of_species!&quot;)
    if (!$method_link_species_set);

# Fetching the query Slice:
my $slice_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor($species, &quot;core&quot;, &quot;Slice&quot;);

throw(&quot;Registry configuration file has no data for connecting to &lt;$species&gt;&quot;)
    if (!$slice_adaptor);

my $query_slice = $slice_adaptor-&gt;fetch_by_region('toplevel', $seq_region, $seq_region_start, $seq_region_end);

throw(&quot;No Slice can be created with coordinates $seq_region:$seq_region_start-&quot;.
    &quot;$seq_region_end&quot;) if (!$query_slice);

# Fetching all the GenomicAlignBlock corresponding to this Slice:
my $genomic_align_block_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    $compara, &quot;compara&quot;, &quot;GenomicAlignBlock&quot; );

my $genomic_align_blocks =
    $genomic_align_block_adaptor-&gt;fetch_all_by_MethodLinkSpeciesSet_Slice(
      $method_link_species_set,
      $query_slice);

my $all_aligns;

# Get a Bio::SimpleAlign object from every GenomicAlignBlock
foreach my $this_genomic_align_block (@$genomic_align_blocks) {
    my $simple_align = $this_genomic_align_block-&gt;get_SimpleAlign;
    push(@$all_aligns, $simple_align);
}

# print all the genomic alignments using a Bio::AlignIO object
my $alignIO = Bio::AlignIO-&gt;newFh(
    -interleaved =&gt; 0,
    -fh =&gt; \*STDOUT,
    -format =&gt; $output_format,
    -idlength =&gt; 10
);

foreach my $this_align (@$all_aligns) {
    print $alignIO $this_align;
}

exit;
</pre>

<h2>Homologies and Protein clusters</h2>

<p>
All the homologies refer to Members. Homology objects store orthologous and paralogous relationships between Members.
</p>


<h3>Member objects</h3>

<p>
A Member represent either a gene or a protein. Most of them
are defined in the corresponding EnsEMBL core database. For
instance, the sequence for the human gene ENSG00000004059
is stored in the human core database.
</p>

<p>
The fetch_by_source_stable_id method of the MemberAdaptor takes two arguments. The first one is the
source of the Member and can be:
</p>

<ul>
<li>ENSEMBLPEP, derived from an EnsEMBL translation</li>
<li>ENSEMBLGENE, derived from an EnsEMBL gene</li>
<li>Uniprot/SWISSPROT, derived from a Uniprot/Swissprot entry</li>
<li>Uniprot/SPTREMBL, derived from a Uniprot/SP-TrEMBL entry</li>
</ul>

<p>
The second argument is the identifier for the Member. Here is a simple example:
</p>

<pre class="code">
# get the MemberAdaptor
my $member_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(
    &quot;pan_homology&quot;,&quot;compara&quot;,&quot;Member&quot;);

# fetch a Member
my $member = $member_adaptor-&gt;fetch_by_source_stable_id(
    &quot;ENSEMBLGENE&quot;,&quot;ENSG00000004059&quot;);

# print out some information about the Member
print $member-&gt;chr_name, &quot; ( &quot;, $member-&gt;chr_start, &quot; - &quot;, $member-&gt;chr_end,
    &quot; ): &quot;, $member-&gt;description, &quot;\n&quot;;
</pre>

<p>
The Member object has several attributes:
</p>
<ul>
<li><code>source_name</code> and <code>stable_id</code> define this Member.</li>
<li><code>chr_name</code>, <code>chr_start</code>, <code>chr_end</code>, <code>chr_strand</code> locate this Member on the genome but
are only available for ENSEMBLGENE and ENSEMBLPEP.</li>
<li><code>taxon_id</code> corresponds to the NCBI taxonomy identifier (see
<a href="http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/">http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/</a>
for more details).</li>
<li><code>taxon</code> returns a Bio::EnsEMBL::Compara::NCBITaxon object.
From this object you can get additional information about the species.</li>
</ul>

<pre class="code">
my $taxon = $member-&gt;taxon;
print &quot;common_name &quot;, $taxon-&gt;common_name,&quot;\n&quot;;
print &quot;genus &quot;, $taxon-&gt;genus,&quot;\n&quot;;
print &quot;species &quot;, $taxon-&gt;species,&quot;\n&quot;;
print &quot;binomial &quot;, $taxon-&gt;binomial,&quot;\n&quot;;
print &quot;classification &quot;, $taxon-&gt;classification,&quot;\n&quot;;
</pre>

<p>
In our example the species is human, so the output will look like this:
</p>
<pre>
common_name:    human
genus:          Homo
species:        sapiens
binomial:       Homo sapiens
classification:	sapiens Homo Hominidae Catarrhini Haplorrhini Primates Euarchontoglires Eutheria Mammalia Euteleostomi Vertebrata Craniata Chordata Metazoa Eukaryota
</pre>


<h3>Homology Objects</h3>

<p>
A Homology object represents either an orthologous or paralogous relationships between two
or more Members.
</p>

<p>
Typically you want to get homologies for a given gene. The HomologyAdaptor has a
fetching method called fetch_all_by_Member(). You will need the Member object for your
query gene, therefore you will fetch the Member first like in this example:
</p>

<pre class="code">
# first you have to get a Member object. In case of homology, it is a gene

my $member_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(&quot;pan_homology&quot;, &quot;compara&quot;, &quot;Member&quot;);
my $member = $member_adaptor-&gt;fetch_by_source_stable_id(&quot;ENSEMBLGENE&quot;,&quot;ENSG00000004059&quot;);

# then you get the homologies where the member is involved

my $homology_adaptor = Bio::EnsEMBL::Registry-&gt;get_adaptor(&quot;pan_homology&quot;, &quot;compara&quot;, &quot;Homology&quot;);
my $homologies = $homology_adaptor-&gt;fetch_all_by_Member($member);

# That will return a reference to an array with all homologies (orthologues in
# other species and paralogues in the same one)
# Then for each homology, you can get all the Members implicated

foreach my $homology (@{$homologies}) {
  # You will find different kind of description
  # UBRH, MBRH, RHS, YoungParalogues
  # see ensembl-compara/docs/docs/schema_doc.html for more details

  print $homology-&gt;description,&quot; &quot;, $homology-&gt;subtype,&quot;\n&quot;;

  # And if they are defined dN and dS related values

  print &quot; dn &quot;, $homology-&gt;dn,&quot;\n&quot;;
  print &quot; ds &quot;, $homology-&gt;ds,&quot;\n&quot;;
  print &quot; dnds_ratio &quot;, $homology-&gt;dnds_ratio,&quot;\n&quot;;
}
</pre>

<p>
Each homology relation has 2 or more members, you should find there the initial member used as a query.
The get_all_MemberAttribute method returns an array of pairs of Member and Attributes. The Member
corresponds to the gene or protein and the Attribute object contains information about how this
Member has been aligned.
</p>

<pre class="code">
my $homology = $homologies-&gt;[0]; # take one of the homologies and look into it

foreach my $member_attribute (@{$homology-&gt;get_all_Member_Attribute}) {

  # for each Member, you get information on the Member specifically and in
  # relation to the homology relation via Attribute object

  my ($member, $attribute) = @{$member_attribute};
  print (join &quot; &quot;, map { $member-&gt;$_ }  qw(stable_id taxon_id)).&quot;\n&quot;;
  print (join &quot; &quot;, map { $attribute-&gt;$_ } qw(perc_id perc_pos perc_cov)).&quot;\n&quot;;

}
</pre>

<p>
You can get the original alignment used to define an homology:
</p>

<pre class="code">
use Bio::AlignIO;

my $simple_align = $homology-&gt;get_SimpleAlign();
my $alignIO = Bio::AlignIO-&gt;newFh(
    -interleaved =&gt; 0,
    -fh =&gt; \*STDOUT,
    -format =&gt; &quot;clustalw&quot;,
    -idlength =&gt; 20);

print $alignIO $simple_align;
</pre>


<h2>Further help</h2>

<p>
For additional information or help mail the <a href="mailto:ensembl-dev&#64;ebi.ac.uk">ensembl-dev</a> mailing list.
You will need to subscribe to this <a href="/info/about/contact/mailing.html">mailing list</a> to use it.
</p>

<!-- End of real content -->

</body>
</html>
