<html>
<head>
<meta name="order" content="2" />
<title>Genome Annotation</title>
</head>

<body>

<h1>Genome Annotation</h1>

<p>The genomes provided by Ensembl Genomes contain annotation on genes and gene function that are 
obtained via import of external data or use of predictive algorithms. This document outlines the 
steps involved in adding annotation to a <a href="/info/genome/assemblies/">genome assembly</a>:</p>
<ol>
  <li>Import protein coding gene models. Ensembl Genomes does not usually carry out primary annotation 
of protein-coding gene models. Gene models are imported either from annotation in 
<a href="/info/genome/annotation/insdc_annotation/">INSDC</a> sequence archive records or from 
other public sources, in which case <a href="/info/genome/annotation/gff_annotation.html">GFF</a> 
is the preferred import format. In addition, Ensembl Genomes is involved in collaborations from 
which <a href="/info/genome/annotation/manual_annotation/">manual annotation</a> is imported. In 2023,
Ensembl Bacteria released gene models generated using <a href="https://academic.oup.com/bioinformatics/article/30/14/2068/2390517">Prokka</a>.</li>
  <li>Annotate <a href="/info/genome/annotation/ncrna.html">non-coding gene models</a></li>
  <li>Annotate <a href="/info/genome/annotation/repeat_features.html">repeat features</a></li>
  <li>Annotate <a href="/info/genome/annotation/protein_features.html">protein features</a></li>
  <li><a href="/info/genome/annotation/cross_references.html">Add cross-references</a> to external data sources</li>
</ol>

</body>
</html>
