<html>
<head>
<meta name="order" content="5" />
<title>RNA gene annotation</title>
</head>

<body>

<h1>RNA gene annotation</h1>

<p>If the genesets that are imported by Ensembl Genomes include RNA genes as well as 
protein-coding genes, these RNA genes will be displayed without modification. However, 
often RNA genes are not provided, or only a subset of biotypes will have been annotated 
(e.g. tRNA genes). In the case of missing or partial RNA gene annotation, Ensembl Genomes 
performs RNA gene prediction.</p>

<p>Currently, we are in a transitional state between two versions of our RNA gene annotation 
pipeline. As of release 36, Ensembl Metazoa uses the "New" method below; the other divisions 
currently use the "Old" method, but are moving to the "New" method in subsequent releases.</p>

<h2>Ensembl Genomes annotation pipeline: New</h2>

<p>The RNA gene annotation uses three sources:</p>

<ul>
  <li><a href="http://www.mirbase.org/">miRBase</a> (release 21) [1] provides genomic locations 
of precursor microRNA genes for a subset of species. These are loaded without any filtering.</li>
  <li><a href="http://lowelab.ucsc.edu/tRNAscan-SE/">tRNAscan-SE</a> (version 1.23) [2] is used 
to predict tRNA genes. tRNA annotation is sometimes difficult in repeat-rich species, leading to 
over-prediction with the default tRNAscan score threshold (=40). We therefore analyse the 
distribution of predicted tRNA genes for each division, to determine a score threshold that 
provides a balance between sensitivity and specificity. For Ensembl Metazoa we use a score 
threshold of 65, and tRNA genes are not annotated if they would overlap either repeat regions 
or protein-coding exons.</li>
  <li><a href="http://rfam.xfam.org/">Rfam</a> (version 12.2) [3] covariance model alignments 
are used to annotate RNA genes of all classes except tRNA. The annotation relies on 
<a href="/info/genome/dna_align/rfam_alignment.html">taxonomically-appropriate alignments</a> 
with <i>cmscan</i>, from the <a href="http://eddylab.org/infernal">Infernal</a> (version 1.1.2) 
software suite [4], which are available as a separate track in the genome browser. Each alignment must pass a set of filters for it to be annotated as an RNA gene:
    <ul>
      <li>Rfam is not just a database of RNA genes; it also contains structures for regulatory 
elements, for example. So non-gene biotypes are filtered out of the potential gene set.</li>
      <li>E-values must be higher than a certain threshold, set on a per-division basis. 
The threshold for Ensembl Metazoa is 1e-6.</li>
      <li>Some RNA sequences are inherently palindromic, due to the base pairing required for 
hairpin structures; this means you often get credible hits on both strands. However, it is usually 
the case that one has a much better E-value, so in the case of overlapping alignments (on either 
the same or different strands), we use the E-value to determine which alignment to convert to a 
gene.</li>
      <li>RNA genes are not annotated if they would overlap either protein-coding exons or 
RNA genes from miRBase.</li>
      <li>It is possible for alignments to partially represent an RNA structure, if there is 
an assembly gap, for example. Such alignments would make partial genes, so are not annotated 
as RNA genes.</li>
      <li>Skewed GC content makes false positives more likely, because the chance of spurious 
base pairing increases. RNA genes are not annotated in areas of high GC or AT content.</li>
    </ul>
  </li>
</ul>

<h2>Ensembl Genomes annotation pipeline: Old</h2>

<p>For all <b>ncRNA</b> except <b>tRNA</b> and <b>rRNA</b> genes, models are predicted by 
aligning a genomic sequence against <a href="http://rfam.sanger.ac.uk/" rel="external">Rfam sequences</a> [5] using <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi" rel="external">BLASTN</a>. 
The BLAST hits are then used to seed <a href="http://infernal.janelia.org" rel="external">Infernal</a>
 searches of the aligned regions with the corresponding Rfam covariance models. The purpose of this 
is to reduce the search space required, as to scan the entire genome with all the Rfam covariance 
models would be extremely CPU-intensive.</p>

<p><b>tRNA</b> is predicted by using <a href="http://selab.janelia.org/tRNAscan-SE">tRNAScan-SE</a> software [2]. Version 1.23 of the program was used, configured for superregnum as appropriate.</p>

<p><b>rRNA</b> is predicted by using <a href="http://www.cbs.dtu.dk/services/RNAmmer">RNAmmer</a> software [6]. Version 1.2 of the program was used, configured for superregnum as appropriate.</p>

<h2>Stable identifiers</h2>

<p>Historically, RNA genes annotated by Ensembl Genomes have had identifiers that encoded the 
division and species name, e.g. EMBMOG00000000027, where "EM" means Ensembl Metazoa, 
"BMO" means <i>Bombyx mori</i>, and "G" indicates a gene (as opposed to "T" for a transcript). 
Starting with Ensembl Metazoa in release 36 (and other divisions in subsequent releases) we are 
switching to a simpler format, applicable across all divisions, with an "ENSRNA" prefix, where 
"ENS" is short for Ensembl, and "RNA" is, hopefully, self-explanatory. E.g. ENSRNA022711053 is 
a gene stable identifier, and transcripts of that gene have a suffix with "T" for transcript 
and a number, e.g. ENSRNA022711053-T1.</p>

<p>Stable identifiers have been mapped, such that <a href="http://metazoa.ensembl.org/Bombyx_mori/Tools/IDMapper?db=core">ID History Converter tool</a> can be used to find new IDs which correspond 
to old IDs (and vice versa). However, note that the new pipeline for RNA gene annotation is 
more rigorous in its filtering than the old pipeline, so old, unsupported genes may have been 
deleted and have no mapping to a new ID.</p>

<h2>References</h2>

<ol>
  <li>
  <p>Kozomara A and Griffiths-Jones S (2014) <a href="http://europepmc.org/abstract/MED/24275495">miRBase: annotating high confidence microRNAs using deep sequencing data</a> <i>Nucl. Acids Res.</i> <b>42</b> D68-73</p>
  </li>
  <li>
  <p>Lowe TM and Eddy SR (1997) <a href="http://europepmc.org/abstract/MED/9023104">tRNAScan-SE: a program for improved detection of transfer RNA genes in genomic sequence</a> <i>Nucl. Acids Res.</i> <b>25</b> 955-964</p>
  </li>
  <li>
  <p>Nawrocki EP <i>et al.</i> (2015) <a href="http://europepmc.org/abstract/MED/25392425">Rfam 12.0: updates to the RNA families database</a> <i>Nucl. Acids Res.</i> <b>43</b> D130-37</p>
  </li>
  <li>
  <p>Nawrocki EP and Eddy SR (2013) <a href="http://europepmc.org/abstract/MED/24008419">Infernal 1.1: 100-fold faster RNA homology searches</a> <i>Bioinformatics</i> <b>29</b> 2933-2935</p>
  </li>
  <li>
  <p>Burge SW <i>et al.</i> (2013) <a href="http://europepmc.org/abstract/MED/23125362">Rfam 11.0: 10 years of RNA families</a> <i>Nucl. Acids Res.</i> <b>41</b> D226-32</p>
  </li>
  <li>
  <p>Lagesen K <i>et al.</i> (2007) <a href="http://europepmc.org/abstract/MED/17452365">RNammer: consistent annotation of rRNA genes in genomic sequences</a> <i>Nucl. Acids Res.</i> <b>35</b> 3100-3108</p>
  </li>
</ol>


</body>
</html>
